Add simple light-client mode (RPC only)

jonasschnelli opened the pull request:
This adds a simple light client mode (RPC only, no wallet support).

With this PR, It is possible to disable auto-request-blocks by passing in `-autorequestblocks=0`.
In that mode, one can request out-of-band blocks by calling `requestblock add ["<blockhash>", ...]`.
Those blocks will then be requested/downloaded and can be loaded via `getblock` (and they will also be passed through ZMQ).

This allows a very simple light-client mode ideally when you already have a validated peer in your trusted network.

This is also a reviewable step towards light client mode for the wallet (which will ultimately allow process separation o the wallet).

jonasschnelli committed:
[RPC] Add requestblocks - a simple way to priorize block downloads

jonasschnelli committed:
Add fAutoRequestBlocks to disabled/enable the verification progress

jonasschnelli committed:
Add -autorequestblocks which then allows to run in non-validation mode (pure light-client mode)

jonasschnelli commented:
Contains overhauled parts from #9483

MIGUELWAXX commented:
cheers

ryanofsky commented:
I'd like to help out here, and I've spent literally days reviewing previous iterations of this change (#9076 #9483 #9171), but I can't figure out if this is going anywhere and if providing more review now is a good use of time.

As I've mentioned previously, I don't think the auxiliary block download class design is great, because it duplicates functionality from the networking layer in a wholly different part of the code, and gives the wallet too much control & responsibility over low level p2p details in the case where it's used to let the wallet code prioritize which blocks to download. I've tried to describe what I think would be better alternatives in my previous review comments, like https://github.com/bitcoin/bitcoin/pull/9483#issuecomment-278799990.

@jonasschnelli, assuming you don't want to take my previous suggestions, and do want to stick with the current design, I think it would be helpful if you could reach out to some other reviewers and get some concept acks for your current approach. It might help to put together some kind of design document, or at least a detailed comment to the top of `auxiliaryblockrequest.h` that lays out how the class works and what your future plans for it are in as much detail as possible, so someone can understand how it's intended to be used without having to wade through all the PRs.

MIGUELWAXX commented:
cheers

jonasschnelli commented:
@ryanofsky:
Thank you very much for the reviews in #9483. I implemented almost all of your suggesting and some of them where really great.
However, I think the current design of having a dedicated class (`CauxiliaryBlockRequest`) makes sense, because...

* An out-of-band (auxiliary) block request in an object, you could have multiple in parallel (assume multiwallet, etc.), in future, we may want to serialise them to disk, etc. Therefore I think it should be capsulated in its own class.

* I think the wallet needs control. For a light-client mode, the wallet is the one that tells the node what it needs. More or less, wallet tells node: "I need block A to E, give them to me as fast as possible", then node may call back "I have block A already ready (on disk), rest will follow soon", etc. Wallet then may say: "Okay, thats enough, you can stop at B already", etc.

* Having the implementation in a designated file makes code overview simpler. It can result in having slightly more lines of code but it should worth it. Having another `main.cpp` for everything that is network related should be avoided now, early enough. I don't think we have a lot of duplicated code, or can you point out what would be more compact if we would implement it directly in `net_processing.cpp`?

* Last but not least, it allows to have a patch-set the requires less rebasing. And this is an important one to me. Eventually, this will be something that "runs along" with the master branch until there has been reasonable amount of testing before it gets eventually merged.

jonasschnelli committed:
Add priority block request queue

jonasschnelli committed:
Don't ActivateBestChain when processing priority block request

jonasschnelli committed:
Optionally allow to skip the toFarAway check in AcceptBlock

jonasschnelli commented:
Followed yesterdays discussion we had in #bitcoin-core-dev. Removed the `CAuxiliaryBlockRequests` and added a `std::map blocksToDownloadFirst`.
Such manually added, priority block downloads will not trigger `ActivateBestChain`.

This PR now also adds a new signal `BlockProcessed()`.

The scope of this PR is not to make the block download interface flexible for multiple "users" (like the validation and eventually the light-client wallet).

jonasschnelli commented:
fixed @ryanofsky points.
This does pass travis now. Thanks in advance for reviews...

ryanofsky reviewed (commented):
Change seems pretty straightforward. Main thing missing is a test for the new RPC call.

It would be useful to get concept ACKs from @gmaxwell and @sipa to make sure concerns from the earlier design discussion (https://botbot.me/freenode/bitcoin-core-dev/msg/88437543/) are addressed now.

I don't have a very clear idea of how this functionality is going to be used, and what future plans for it are, so I personally wouldn't mind seeing a list of what the next steps & future prs might look like, or knowing some applications that could use the new option & RPC. But probably these things are clearer to others.

jonasschnelli commented:
Thanks @ryanofsky. Agree about required conceptual ACKs from other devs.
The long term goal was sketched in #9483, basically a light client mode for Bitcoin-Core that would allow node/wallet separation.

jonasschnelli committed:
[QA] add requestblocks test

ryanofsky reviewed (commented):
utACK 1794a11037150b0d2c6ba2eb9f06e310f51f7c1c. Change looks great: code & test are straightforward and commit history is well thought out.

I still can't figure out if new RPC is actually supposed to be useful for something other than debugging/testing. Seems fine if it isn't, though, because it's a simple wrapper around the functions for downloading blocks of order, which obviously are useful for SPV.

jonasschnelli commented:
Overhauled and fixed @ryanofsky points (mostly nits), also removed the new (unused) signal.

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -1532,6 +1537,63 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
+        return ret;
+    }
+    else if (request.params[0].get_str() == "add") {
+        if (request.params.size() < 2) {
```
In commit "[RPC] Add requestblocks - a simple way to priorize block downloads"

Don't really need this check, because the get_array call below will trigger it's own error about the value not being an array. But if you prefer this error, you should change the check to request.params[1].isNull() to avoid making a distinction between a null and a missing value (#10281, #10783).

ryanofsky reviewed (commented):
utACK 9723672ced8b48eeee4847e6c0e0ed39833c7eaf. Changes since last review were dropping a commit that added a `BlockProcessed` signal, changing order of other commits, changing brace styles, and making various suggested tweaks.

jonasschnelli commented:
Overhauled first commit to make sure we request blocks in order:
https://github.com/bitcoin/bitcoin/pull/10794/commits/5927f7f1d6c7eb1bc7fb7a39dfda277ad5e4474c#diff-eff7adeaec73a769788bb78858815c91R483

jonasschnelli committed:
Add new ProcessPriorityRequest main signal

jonasschnelli committed:
Process priority requests

jonasschnelli committed:
Pass priority requests through ZMQ notifications

jonasschnelli commented:
* Overhauled once again. Added the processing logic to ensure blocks are processed in order (makes it much simpler to process by "the other side").
* Added a new main signal for the processing (reusing BlockChecked seems wrong).
* Priority requests are now pushed through signal

This is now tested on my SPV branch and could be the first reviewable step towards light clients / wallet process separation.

jtimon commented:
Needs rebase

nopara73 commented:
@ryanofsky Reflecting to your usefullness concern:  
I wrote a full block downloading SPV wallet ([HiddenWallet](https://github.com/nopara73/HiddenWallet/)), the main reason is, because without such architecture [private transaction retrieval is hard](https://groups.google.com/forum/#!msg/bitcoinj/Ys13qkTwcNg/9qxnhwnkeoIJ), if not impossible.  
When some form of light functionality gets into Core I would personally consider porting HiddenWallet to on top of Core, since I cannot compete with Core's speed and stability.

ryanofsky commented:
That's really interesting. @nopara73, if I understand you correctly, you are saying that if the `requestblocks` JSON-RPC API from this PR is added to bitcoin core, then [HiddenWallet](https://github.com/nopara73/HiddenWallet/) would be able to call it to download blocks instead of depending on [NBitcoin](https://github.com/MetacoSA/NBitcoin)?

nopara73 commented:
@ryanofsky 

> JSON-RPC API from this PR is added to bitcoin core, then HiddenWallet would be able to call it to download blocks instead of depending on NBitcoin?  

Yes.

jonasschnelli commented:
Rebased

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1587,6 +1592,60 @@ UniValue savemempool(const JSONRPCRequest& request)
 
+UniValue requestblocks(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 3)
```
Should be `> 2`?

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -645,6 +649,7 @@ UniValue getblockheader(const JSONRPCRequest& request)
             "{\n"
             "  \"hash\" : \"hash\",     (string) the block hash (same as provided)\n"
             "  \"confirmations\" : n,   (numeric) The number of confirmations, or -1 if the block is not on the main chain\n"
+            "  \"validated\" : n,       (boolean) True if the block has been validated\n"
```
Should update `getblockheader` test?
https://github.com/bitcoin/bitcoin/blob/84fa64598e8b7d6901c546e938fb0eeda625a8e8/test/functional/blockchain.py#L173-L198

jonasschnelli replied:
done

promag commented on src/net_processing.h:
```diff
@@ -81,4 +83,18 @@ bool GetNodeStateStats(NodeId nodeid, CNodeStateStats &stats);
+ * Blocks requested with priority will be downloaded and processed first
+ * Downloaded blocks will not trigger ActivateBestChain
+ */
+void AddPriorityDownload(std::vector<const CBlockIndex*>& blocksToDownload);
```
Arg reference should be const:
```cpp
const std::vector<const CBlockIndex*>& blocks_to_download
```

promag commented on src/net_processing.cpp:
```diff
@@ -3647,6 +3699,74 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+}
+
+size_t CountPriorityDownloads() {
+    // return a copy
```
Copy of?

promag commented on src/net_processing.cpp:
```diff
@@ -3647,6 +3699,74 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+            currentBlock = std::make_shared<const CBlock>(loadBlock);
+        }
+        else {
+            break;
```
Add comment explaining this case?

promag commented on src/net_processing.cpp:
```diff
@@ -3647,6 +3699,74 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+}
+
+void ProcessPriorityRequests(const std::shared_ptr<CBlock> blockRef) {
+    LOCK(cs_main);
```
`cs_main` is locked throughout. It is enough to lock when dequeuing right? If dequeuing fails then return.

promag commented on src/net_processing.cpp:
```diff
@@ -3647,6 +3699,74 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+
+        // allow processing through signal
+        GetMainSignals().ProcessPriorityRequest(currentBlock, r.pindex);
+        LogPrint(BCLog::NET, "process priority block request (%s) height=%d\n", r.pindex->GetBlockHash().ToString(), r.pindex->nHeight);
```
Move log before signal?

jonasschnelli replied:
I think it's okay after, but should say "processed" instead of "process".

promag commented on src/net_processing.cpp:
```diff
@@ -3647,6 +3699,74 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+        LogPrint(BCLog::NET, "process priority block request (%s) height=%d\n", r.pindex->GetBlockHash().ToString(), r.pindex->nHeight);
+
+        // remove processed block from queue
+        it = blocksToDownloadFirst.erase(std::remove_if(blocksToDownloadFirst.begin(), blocksToDownloadFirst.end(), [&r](const PriorityBlockRequest &rB) {
```
Is removing duplicates here really necessary? Blocks with `BLOCK_HAVE_DATA` will be skipped.

jonasschnelli replied:
With the current implementation, `blocksToDownloadFirst` can have multiple of the same block. I'd say it should remain here...

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1587,6 +1592,60 @@ UniValue savemempool(const JSONRPCRequest& request)
+            "2. blockhashes       (array, optional) the hashes of the blocks to download\n"
+            "\nResult:\n"
+            "   add: <null>\n"
+            "   flush: <true|false> (if the the queue wasn't empty)\n"
```
What is the purpose of the return value?

jonasschnelli replied:
You can determine if you have flushed something or not.

promag commented on src/net_processing.h:
```diff
@@ -81,4 +83,18 @@ bool GetNodeStateStats(NodeId nodeid, CNodeStateStats &stats);
+void ProcessPriorityRequests(const std::shared_ptr<CBlock> block);
+bool FlushPriorityDownloads();
+size_t CountPriorityDownloads();
+void ProcessPriorityRequests(const std::shared_ptr<CBlock> block);
```
Remove, duplicate declaration.

promag commented on src/net_processing.cpp:
```diff
@@ -473,9 +498,26 @@ void FindNextBlocksToDownload(NodeId nodeid, unsigned int count, std::vector<con
+    if (!blocksToDownloadFirst.empty()) {
+        for (const PriorityBlockRequest &r: blocksToDownloadFirst) {
+            if (r.downloaded) continue;
+            if (r.pindex && state->pindexBestKnownBlock != NULL && state->pindexBestKnownBlock->nHeight >= r.pindex->nHeight && !mapBlocksInFlight.count(r.pindex->GetBlockHash())) {
```
Split in multiple tests. Suggestion for loop body:
```cpp
if (!r.pindex) continue;
if (state->pindexBestKnownBlock == nullptr) continue;
if (state->pindexBestKnownBlock->nHeight < r.pindex->nHeight) continue;
if (mapBlocksInFlight.count(r.pindex->GetBlockHash())) continue;

vBlocks.push_back(r.pindex);

if (vBlocks.size() == count) break;
```
IMHO more readable and more easier to add comments.

promag replied:
Otherwise, replace `NULL` with nullptr and use `break` instead of `return true`.

promag reviewed (commented):
LGTM, will play a bit, meanwhile some comments below.

Test that flush empties the queue:
 - disconnect node
 - requestblocks add ["valid hash"]
 - requestblocks flush
 - requestblocks status
 - assert count equal 0

Test that no block is enqueued if one is unknown:
 - requestblocks add ["valid hash", "invalid hash"]
 - assert the above error
 - requestblocks status
 - assert count equal 0

promag commented on src/net_processing.cpp:
```diff
@@ -136,6 +139,13 @@ namespace {
     /** Expiration-time ordered list of (expire time, relay map entry) pairs, protected by cs_main). */
     std::deque<std::pair<int64_t, MapRelay::iterator>> vRelayExpiration;
+
+    struct PriorityBlockRequest {
```
Same for other structs.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1587,6 +1592,60 @@ UniValue savemempool(const JSONRPCRequest& request)
+            );
+
+    if (request.params[0].get_str() == "flush") {
+        return UniValue(FlushPriorityDownloads());
```
Assert `request.params[1]` is null. Same for status.

nopara73 commented:
# Simple RPC App Example

Just to be sure I understand correctly this PR and the application I am considering will work as I expect to work:

How would the wallet behave? Best effort or disabled?  

So let's say is it possible to write a software, that uses Bitcoin Core's wallet with autorequestblocks=0 in the following way:  

1. Sync up the header.  
2. Create a wallet, save the creation time (or block height).  
3. Subscribe for new header notifications.  
4. Every time a new header comes call `requestblock add blockhash`

So the wallet, from the information it has available executes the rpc commands properly based on best effort, right? And is not supposed to fail.  

This gives me the possibility to implement leapfrogging for example. (Enable the user to say: hey, I didn't do or wait for any transaction in the last 3 months and I have no intention to sync it up, it's ok if I get the blocks from now on.)  

# Other Concerns

1. Since this mode cannot have up to date utxo set, how do you broadcast transactions? How do you make sure it's valid?  
2. How do you accept transactions to your mempool without making sure it spends an utxo? Are you also propagating them?  
3. What would getblockcount return?

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -1587,6 +1592,60 @@ UniValue savemempool(const JSONRPCRequest& request)
+    if (request.params[0].get_str() == "flush") {
+        return UniValue(FlushPriorityDownloads());
+    }
+    else if (request.params[0].get_str() == "status") {
```
Why?

nopara73 commented:
@jonasschnelli I'd like to make the last couple of things clear before I compile and test it, just to know what to expect.  

# 1

> there is a commit available in #9483 7ca1a87 that does allow broadcasting transactions without an utxo set.

Meaning, `sendrawtransaction` will always return null?

# 2

> This PR does not touch the wallet (that will come later, would be to large of a PR).  

Meaning, (1) disables the wallet, (2) makes the wallet rpcs return sometimes incorrect results, (3) you don't know what the rpcs would return?

jonasschnelli commented:
> Meaning, sendrawtransaction will always return null?

No. It will return the txid (as expected). But not in this PR (that would be with something like https://github.com/bitcoin/bitcoin/commit/7ca1a8738a878078c0d1546f7743a05ee474dd1b).

> Meaning, (1) disables the wallet, (2) makes the wallet rpcs return sometimes incorrect results, (3) you don't know what the rpcs would return?

The wallet runs the same way as in master.
All wallet commands do work exactly the same way as the do in standard IBD mode.
Blocks processed with the new `requestblocks` will be ignored by the wallet (otherwise this PR would be too large).

The scope of this PR is to add the necessary network processing infrastructure to later add an SPV mode to the wallet.

ryanofsky commented:
>> It would be useful to get concept ACKs from @gmaxwell and @sipa to make sure concerns from the earlier design discussion (https://botbot.me/freenode/bitcoin-core-dev/msg/88437543/) are addressed now.

> Agree about required conceptual ACKs from other devs.

Last I looked code was in good shape, but I don't think there was ever more design feedback on this. Maybe it should be brought up at an IRC meeting?

nopara73 commented:
This will be especially useful when combined with BIP158 (Compact Block Filters for Light Clients.) 
https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki

I imagine, based on this PR someone will eventually make a Bitcoin Core light client after Core starts to serve filters. At least that'll be a low hanging fruit.

Sjors commented:
Though I'm not sure what it would do on its own, perhaps the `PriorityBlockRequest` could be a separate PR?

Same for `requestblocks` RPC; could that be useful in pruned nodes to retrieve a pruned block?

dmoney0546 commented:
This is dmarcus Eason and we need to talk about this because Daniel have it
right and I need it fast

On Feb 20, 2018 08:51, "Sjors Provoost" <notifications@github.com> wrote:

> Though I'm not sure what it would do on its own, perhaps the
> PriorityBlockRequest could be a separate PR?
>
> Same for requestblocks RPC; could that be useful in pruned nodes to
> retrieve a pruned block?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/10794#issuecomment-367001197>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ai9_b5He-pOevv4ILEYCDQqRhqi_MHKSks5tWtwAgaJpZM4OUX1v>
> .
>

jonasschnelli commented:
Rebased

Sjors commented:
I rebased it here: https://github.com/Sjors/bitcoin/commits/2018/08/spv-rpc

@ryanofsky said:
> I still can't figure out if new RPC is actually supposed to be useful for something other than debugging/testing

The bigger picture for me is #9483, the ability for users to get started while IBD happens in the background.

However, a use case I have in mind for `requestblocks add` is https://github.com/ElementsProject/lightning/issues/1297, in particular the ability for a lightning client to re-download historic blocks from a pruned node. This actually works now.

`-autorequestblocks` is useful when running on a phone (iOs: #11720, Android: [ABCore](https://github.com/greenaddress/abcore)). A native application could use the RPC and only allow fetching blocks when on wifi. Being able to toggle automatic download via RPC would be nice, but can wait.

Being able to fetch blocks by height range (or up to a certain height) instead of hashes would be nice, but that's something that can be built on top of this.

I noticed there's no log entry when you `requestblocks add`.

Can you explain when `priorityRequest` is `true`?

Did some light testing and couldn't find problems, but more people should try :-)

Sjors commented on test/functional/rpc_requestblocks.py:
```diff
@@ -0,0 +1,62 @@
+        assert_equal(block['validated'], False)
+
+        #prevblock must not be available
+        assert_raises_rpc_error(-1, "Block not found on disk", self.nodes[1].getblock, block['previousblockhash'])
```
Maybe do a prune here and then refetch the old block?

practicalswift commented on src/rpc/blockchain.cpp:
```diff
@@ -1626,6 +1630,60 @@ UniValue savemempool(const JSONRPCRequest& request)
+        return NullUniValue;
+    }
+    else {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Unkown action");
```
Typo found by `codespell`: Unkown
